### PR TITLE
Issue49 take2

### PR DESCRIPTION
**Pull request by goodletb**
# Ship Movement:
- After two enemy ships are destroyed, the player gains the ability to move their ships.
- Ships can move North, South, East, and West with North being the top of the board.
- All squares on the board other than ships (attacks, misses, etc) will remain in the same place.
- Player gets two moves per game.
- Any ships that cannot move will stay put, all others will move.
- Ships cannot move if they are on the edge, or another ship is blocking them.
- Submarines can only move under other ships if they are submerged, otherwise the ship will block them.

# Other Fixes:
- fixed issue #58 , players will now unlock the space laser after the appropriate amount of ships are destroyed.

### CODE REVIEW

**Comment by goodletb** [comment]
# This PR implements #49 , and fixes #58

**Comment by barnesjt** [comment]
Awesome, I'm gonna do a full review, but here's the test coverage:
Selection_024

Over 80% line coverage in the models package, so we're good.

File: src/main/java/assets/game.js [hunk 1]
```diff
@@ -25,6 +27,20 @@ function makeGrid(table, isPlayer, gridSize) {
     }
 }
 
+function canMoveCheck()
+{
```
Comment (barnesjt) [inline src/main/java/assets/game.js:31]: Stylistically, I think we are avoiding putting open braces on new lines.

File: src/main/java/assets/game.js [hunk 2]
```diff
@@ -25,6 +27,20 @@ function makeGrid(table, isPlayer, gridSize) {
     }
 }
 
+function canMoveCheck()
```
Comment (barnesjt) [inline src/main/java/assets/game.js:30]: I like breaking this out into it's own function, this is something I know I should've done more.

File: src/main/java/assets/game.js [hunk 3]
```diff
@@ -81,7 +97,7 @@ function markHits(board, elementId, surrenderText) {
 
 function redrawGrid() {
 
-    if(spaceLaserEngaged == false && game.opponentsBoard.remainingShips < 3 && !isSetup) {
+    if(spaceLaserEngaged == false && game.opponentsBoard.remainingShips < 4 && !isSetup) {
```
Comment (barnesjt) [inline src/main/java/assets/game.js:100]: Thanks for fixing this, the space laser needing to engage after 1 sunk ship (and not 2) was an oversight on my part.

File: src/main/java/assets/game.js [hunk 4]
```diff
@@ -318,6 +335,38 @@ function place(size) {
     }
 }
 
+function killTwoShips()
```
Comment (barnesjt) [inline src/main/java/assets/game.js:338]: Ooh, I see you must've added this for testing. I've got not problem with this, I played literally dozens and dozens of games just to test features I put in. Smart!

File: src/main/java/controllers/MovementGameAction.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    @JsonProperty private Game game;
+    @JsonProperty private String direction;
+
+    public Game getGame() {
+        return game;
+    }
+
+    public void replaceSubmarine()
+    {
+       for(int i = 0; i < this.getGame().getPlayersBoard().getShips().size(); i++)
```
Comment (barnesjt) [inline src/main/java/controllers/MovementGameAction.java:19]: Yeah you were right, this ended up really ugly. I had problems accessing the submarine how I wanted to as well. I think if we were continuing to develop this, we could refactor quite a bit.

In fact, I believe that Sprint 3 should've just been refactoring. Adding features alongside refactoring while having a requirement to not modify many function signatures really tied our hands behind our back...

File: src/main/java/cs361/battleships/models/Board.java [hunk 6]
```diff
@@ -36,6 +37,131 @@ public boolean placeShip(Ship ship, int x, char y, boolean isVertical) {
 		} else return false;
 	}
 
+	public void moveShips(String direction)
```
Comment (barnesjt) [inline src/main/java/cs361/battleships/models/Board.java:40]: All this looks pretty good, it's readable and has logic broken into a different function. Looks pretty good and appears to work correctly.

File: src/main/java/cs361/battleships/models/Board.java [hunk 7]
```diff
@@ -36,6 +37,131 @@ public boolean placeShip(Ship ship, int x, char y, boolean isVertical) {
...
+				return false;
+			}
+			if(square.getRow() > 10 || square.getRow() < 1)
+			{
+				System.out.println("FAILING ROW CHECK");
+				return false;
+			}
+		}
+
+		for(var propSq : proposedMove)
```
Comment (barnesjt) [inline src/main/java/cs361/battleships/models/Board.java:84]: This gets pretty ugly with the stacks loops and multiple ifs.

File: src/main/java/cs361/battleships/models/Board.java [hunk 8]
```diff
@@ -36,6 +37,131 @@ public boolean placeShip(Ship ship, int x, char y, boolean isVertical) {
...
+					}
+				}
+			}
+		}
+
+
+		return true;
+	}
+
+	public Ship getDirectionmostShip(String direction)
```
Comment (barnesjt) [inline src/main/java/cs361/battleships/models/Board.java:108]: Whats with most being uncapitalized?

File: src/main/java/cs361/battleships/models/Board.java [hunk 9]
```diff
@@ -36,6 +37,131 @@ public boolean placeShip(Ship ship, int x, char y, boolean isVertical) {
...
+	{
+		if(this.getShips().size() == 0)
```
Comment (barnesjt) [inline src/main/java/cs361/battleships/models/Board.java:110]: I think stylistically I like it more condensed, but we don't really have an agreed upon style actually. Also I noticed comments are missing here. I do think it reads just fine regardless, but I wonder if there was a more concise way to do the same thing (maybe not).

File: src/main/java/cs361/battleships/models/Board.java [hunk 10]
```diff
@@ -104,7 +230,7 @@ else if((x>10 || x<0) || (y > 'J' || y < 'A')){
 			res.setResult(AttackStatus.MISS);
 		}
 		//Setting laser for the next attack cycle
-		if(remainingShips < 3) laserEnabled = true;
+		if(remainingShips < 4) laserEnabled = true;
```
Comment (barnesjt) [inline src/main/java/cs361/battleships/models/Board.java:233]: Again, thanks for fixing this fault I left in.

File: src/main/java/cs361/battleships/models/Ship.java [hunk 11]
```diff
@@ -87,4 +119,30 @@ public boolean isSunk() {
...
 		return this.sunk;
 	}
+
+	public void move(String direction)
+	{
+		System.out.println(String.format("MOVING %s %s", this.getKind(), direction));
+		for(var square : this.getOccupiedSquares())
+		{
+			switch(direction)
+			{
```
Comment (barnesjt) [inline src/main/java/cs361/battleships/models/Ship.java:129]: Oh yeah, a switch statement, could've maybe used this above?

File: src/main/java/cs361/battleships/models/Submarine.java [hunk 12]
```diff
@@ -16,6 +17,17 @@ public Submarine(String kind) {
 		this.kind = kind.toUpperCase();
 	}
 
+	public Submarine(Ship ship)
```
Comment (barnesjt) [inline src/main/java/cs361/battleships/models/Submarine.java:20]: It's a shame to have to do it this way, but I understand the need, and we did need to get this done. Don't feel too bad about getting it to work. If we had more time I would totally like fixing this up.

File: src/main/java/views/ApplicationController/index.ftl.html [hunk 13]
```diff
@@ -22,6 +22,12 @@ <h2 id="player_prompt">Place your Ships</h2>
 <div id="visibleButtons">
     <button id="new_game">New Game</button>
     <button id="surrender">Surrender</button>
+    <div id="movement_buttons" class="hidden">
```
Comment (barnesjt) [inline src/main/java/views/ApplicationController/index.ftl.html:25]: I would like to see this interface spruced up a bit, but our UI prototypes did not have considerations for a bunch more buttons. If we were going forward, we'd definitely come up with some new design ideas.

File: src/test/java/cs361/battleships/models/BoardTest.java [hunk 14]
```diff
@@ -206,5 +206,70 @@ public void testSubmergedSubmarinePlacement(){
         assertTrue(board.placeShip(new Submarine("SUBMARINE", true),5,'B', false));
 
     }
+
```
Comment (barnesjt) [inline src/test/java/cs361/battleships/models/BoardTest.java:209]: Really good tests you've written here. If you haven't already make sure you detail the confirmation tests you did for your user story on the wiki page.

**Review by barnesjt — APPROVED** [review]
Great job here, you had a lot of work to do, including editing files we haven't had to touch yet. 

Unfortunately, I think we were victims of the lack of refactoring we were able to accomplish in Sprint 3 along-side features. Having this dual-purpose in the same sprint, where refactoring needed to be done before any feature work caused the refactoring process to be short-changed.

I think you've done a great job here. The fact that the submarine features and ship movement features were successfully developed independently in the same sprint shows there is at least some method to the madness that is our project.

I approve this Pull-Request and would merge. However I would like to give our other two team members a chance to officially review before this is merged, tagged with sprint-4, and submitted.

**Review by ktran0918 — COMMENTED** [review]
A couple of tests reveal movement to work correctly. My tip would be to comment your code, for example where a function call is part of a chain of function calls of object instances:

`this.getGame().getPlayersBoard().getShips().set(i, sub);`

Careful reading should reveal that this sets the ship to a submarine in the i-th order in a list of ships, but it takes some time. This adds "figuring out" to the review process; i.e., instead of spending time to see whether your code matches your intentions, the reviewer has to spend time understanding your code, thus taking away from their effort to spot errors or think of better solutions.

**Pull request merged**